iquip/apps/oven.py

```diff
@@ -0,0 +1,521 @@
+        displayLayout.addWidget(self.voltageDisplayBox)
+        displayLayout.addWidget(self.timerDisplayBox)
+        displayLayout.addWidget(self.timerResetButton)
+        inputLayout = QHBoxLayout()
+        inputLayout.addWidget(self.currentInputBox)
+        inputLayout.addWidget(self.timerInputBox)
+        inputLayout.addWidget(self.outputButton)
+        innerLayout = QVBoxLayout(self._inner)
+        innerLayout.addLayout(displayLayout)
+        innerLayout.addLayout(inputLayout)
+        layout = QVBoxLayout(self)
+        layout.addWidget(self.connectionButton)
+        layout.addWidget(self._inner)
+        # signal connection
+        self.connectionButton.clicked.connect(self._connectionButtonClicked)
+        self.outputButton.clicked.connect(self._outputButtonClicked)
+        # initialize state
+        self.setConnected(False)
+
+    @pyqtSlot(bool)
+    def setConnected(self, open_: bool):
+        """Sets the current connection status.
+
+        This also changes the enabled status and the connection button text.
+        
+        Args:
+            open_: True for open, False for closed.
+        """
+        self._inner.setEnabled(open_)
+        self.connectionButton.setEnabled(True)
+        self.connectionButton.setText("Close" if open_ else "Open")
+
+    def isConnected(self) -> bool:
+        """Returns whether the client is currently connected."""
+        return self.connectionButton.isChecked()
+
+    @pyqtSlot(bool)
+    def _connectionButtonClicked(self, checked: bool):
+        """Connection button is clicked.
+        
+        Args:
+            checked: True for opening the target, False for closing.
+        """
+        self.connectionButton.setEnabled(False)
+        if checked:
+            self.openTarget.emit()
+        else:
+            self.closeTarget.emit()
+
+    @pyqtSlot(float)
+    def setDisplayedCurrent(self, current: float):
+        """Sets the current displayed on the widget.
+        
+        Args:
+            current: Current.
+        """
+        self.currentDisplayBox.setValue(current)
+
+    @pyqtSlot(float)
+    def setDisplayedVoltage(self, voltage: float):
+        """Sets the voltage displayed on the widget.
+        
+        Args:
+            voltage: Voltage.
+        """
+        self.voltageDisplayBox.setValue(voltage)
+
+    @pyqtSlot(float)
+    def setDisplayedTimer(self, on_duration: float):
+        """Sets the timer displayed on the widget.
+        
+        Args:
+            on_duration: Duration while the output is turned on.
+        """
+        self.timerDisplayBox.setValue(on_duration)
+
+    @pyqtSlot(bool)
+    def setOutput(self, output_: bool):
+        """Sets the current output status.
+
+        This also changes the enabled status and the output button text.
+        
+        Args:
+            open_: True for turned on, False for turned off.
+        """
+        self.outputButton.setEnabled(True)
```

Comment:
Why is this necessary?

Reply:
This is because when the button is clicked, it is disabled in `_outputButtonClicked()` to prevent from that one clicks the button before the output is applied.